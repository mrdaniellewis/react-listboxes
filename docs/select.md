# Select

Generates an HTML `<select>`, but with a more convenient way of setting the options and responding to changed values.

This uses the same options as `<DropDown>` and `<ComboBox>` and maybe used interchangeably.

## Warning

Many users find `<select>` controls difficult to use.  A radio group may be more appropriate.

## Usage

This is a stateless controlled component.  You must respond to `onValue` or `onChange` to update the selected value.

```js
const [value, setValue] = useState(initialValue);

<Select
  options={options}
  value={value}
  onValue={setValue}
/>
```

| Prop               | Type       | Purpose                                                   |
| ----               | ----       | ----                                                      |
| `blank`            | `String`   | Set a placeholder option                                  |
| `options`          | `Array`    | The set of options.  See options.                         |
| `value`            | `Any`      | The currently selected option                             |
| `onValue`          | `Function` | Callback will be called with the selected option onChange |
| `mapOption`        | `Function` | Use to map options. See options                           |
| `optionProps`      | `Object`   | Props to add to all `<option>` elements                   |
| `optGroupProps`    | `Object`   | Props to add to all `<optgroup>` elements                 |
| `ref`              |            | Will be passed to `<select>` element                      |
| Any other property |            | Will be added to the `<select>` element                   |

### Options

Options is an array of either:

- `String`
- `Number`
- `null` or `undefined` - will be treated as an empty string
- an object with the following properties:

| Prop               | Type      | Purpose                                              |
| ----               | ----      | ----                                                 |
| `label`            | `String`  | The label of the option (required)                   |
| `disabled`         | `Boolean` | Is the option disabled                               |
| `group`            | `String`  | Label to group options under                         |
| `value`            | `Object`  | Object value used to compare options                 |
| `id`               | `Object`  | Fallback value used to compare options               |
| `html`             | `Object`  | Additional html attributes to be added to the option |
| Any other property |           | Ignored                                              |

When an option is selected `onValue` will be called with the selected option.

When determining which option is selected the option and `value` is compared
by converting to a string value using the equivalent of
`String(option?.value ?? option?.id ?? option?.label ?? option ?? '')`.

If your option does not match the above signature, you can use `mapOption` to match the signature.

```js
const [value, setValue] = useState(initialValue);

const mapOption = useCallback(({ name, deleted }) => {
  return {
    label: name,
    disabled: deleted,
  };
}, []);

<Select
  options={options}
  value={value}
  setValue={setValue}
  mapOption={mapOption}
/>
```
