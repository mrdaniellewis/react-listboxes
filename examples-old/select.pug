extends layout.pug

block variables
  - const title = 'Select'

block main
  h1
    code
      :escape
        <Select />

  section
    h2 Select with string options

    div(data-example='select/string_options')

    details
      summary Code
      pre
        code.language-js
          include:escape:fakeInclude select/string_options.jsx

  section
    h2 Select with blank options

    div(data-example='select/blank_option')

    details
      summary Code
      pre
        code.language-js
          include:escape:fakeInclude select/blank_option.jsx

  section
    h2 Select with object options

    div(data-example='select/object_options')

    details
      summary Code
      pre
        code.language-js
          include:escape:fakeInclude select/object_options.jsx

  section
    h2 Select with disabled option

    div(data-example='select/disabled_option')

    details
      summary Code
      pre
        code.language-js
          include:escape:fakeInclude select/disabled_option.jsx

  section
    h2 Select with groups

  section
    h2 Select with custom classes
