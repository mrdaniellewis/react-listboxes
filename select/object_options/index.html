<html lang="en-GB"><head><meta charset="utf-8"><title>Select with object options</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" href="/react-listboxes/styles.12a99a2f.css"><script src="https://polyfill.io/v3/polyfill.min.js?flags=gated&amp;features=default%2Ces2015%2Ces2016%2Ces2017" crossorigin="anonymous"></script></head><body><nav><div>React combo boxes</div><details><summary>Select</summary><a href="/react-listboxes/select.html">About</a><a href="/react-listboxes/select/string_options/index.html">String options</a><a>Object options</a><a href="/react-listboxes/select/placeholder_option/index.html">Placeholder option</a><a href="/react-listboxes/select/disabled_option/index.html">Disabled option</a><a href="/react-listboxes/select/grouped_options/index.html">Grouped options</a><a href="/react-listboxes/select/mapped_options/index.html">Mapped options</a><a href="/react-listboxes/select/html_options/index.html">HTML options</a></details><details><summary>Drop down</summary><a href="/react-listboxes/drop_down.html">About</a><a href="/react-listboxes/drop_down/string_options/index.html">String options</a><a href="/react-listboxes/drop_down/object_options/index.html">Object options</a><a href="/react-listboxes/drop_down/placeholder_option/index.html">Placeholder option</a><a href="/react-listboxes/drop_down/disabled_option/index.html">Disabled option</a><a href="/react-listboxes/drop_down/disabled/index.html">Disabled drop down</a><a href="/react-listboxes/drop_down/grouped_options/index.html">Grouped options</a><a href="/react-listboxes/drop_down/html_options/index.html">HTML options</a><a href="/react-listboxes/drop_down/mapped_options/index.html">Mapped options</a><a href="/react-listboxes/drop_down/without_managed_focus/index.html">Without managed focus</a><a href="/react-listboxes/drop_down/aria_1_2_groups/index.html">With ARIA 1.2 groups</a><details><summary>Edge cases</summary><a href="/react-listboxes/drop_down/identical_options/index.html">Identical options</a><a href="/react-listboxes/drop_down/no_options/index.html">No options</a><a href="/react-listboxes/drop_down/all_disabled_options/index.html">All disabled options</a></details></details><details><summary>Combo box</summary><a href="/react-listboxes/combo_box.html">About</a><a href="/react-listboxes/combo_box/string_options/index.html">String options</a><a href="/react-listboxes/combo_box/object_options/index.html">Object options</a><a href="/react-listboxes/combo_box/disabled_option/index.html">Disabled option</a><a href="/react-listboxes/combo_box/grouped_options/index.html">Grouped options</a><a href="/react-listboxes/combo_box/html_options/index.html">HTML options</a><a href="/react-listboxes/combo_box/mapped_options/index.html">Mapped options</a><a href="/react-listboxes/combo_box/without_autocomplete/index.html">Without autocomplete</a><a href="/react-listboxes/combo_box/without_managed_focus/index.html">Without managed focus</a><a href="/react-listboxes/combo_box/inline_autocomplete/index.html">With inline autocomplete</a><a href="/react-listboxes/combo_box/aria_1_0/index.html">ARIA 1.0</a><a href="/react-listboxes/combo_box/aria_1_1/index.html">ARIA 1.1</a><a href="/react-listboxes/combo_box/aria_1_2_groups/index.html">ARIA 1.2 groups</a><a href="/react-listboxes/combo_box/listbox_footer/index.html">Listbox footer</a><details><summary>Searches</summary><a href="/react-listboxes/searchers.html">About</a><a href="/react-listboxes/combo_box/async_search/index.html">Async search</a><a href="/react-listboxes/combo_box/search_suggestions/index.html">Search suggestions</a><a href="/react-listboxes/combo_box/background_autocomplete/index.html">With background autocomplete</a></details><details><summary>Edge cases</summary><a href="/react-listboxes/combo_box/identical_options/index.html">Identical options</a><a href="/react-listboxes/combo_box/all_disabled_options/index.html">All disabled options</a><a href="/react-listboxes/combo_box/autoselect_disabled_option/index.html">Disabled option and autoselect</a></details></details><details><summary>Table layout</summary><a href="/react-listboxes/list_box_table.html">About</a><details><summary>Drop down</summary><a href="/react-listboxes/list_box_table/drop_down/index.html">Table</a><a href="/react-listboxes/list_box_table/drop_down_headers/index.html">With headers</a><a href="/react-listboxes/list_box_table/drop_down_grouped/index.html">Grouped</a></details><details><summary>Combo box</summary><a href="/react-listboxes/list_box_table/combo_box/index.html">Table</a><a href="/react-listboxes/list_box_table/combo_box_headers/index.html">With headers</a><a href="/react-listboxes/list_box_table/combo_box_grouped/index.html">Grouped</a></details></details></nav><script src="/react-listboxes/side_collapse.da5ebe24.js"></script><main><h1 id="select-with-object-options">Select with object options</h1>
<p>A select whose options are an array of objects.</p>
<div data-react-example="select/object_options"></div><pre class="language-js"><code class="language-js"><span class="token keyword">import</span> React<span class="token punctuation">,</span> <span class="token punctuation">{</span> useState <span class="token punctuation">}</span> <span class="token keyword">from</span> <span class="token string">'react'</span><span class="token punctuation">;</span>
<span class="token keyword">import</span> <span class="token punctuation">{</span> Select <span class="token punctuation">}</span> <span class="token keyword">from</span> <span class="token string">'@citizensadvice/react-combo-boxes'</span><span class="token punctuation">;</span>

<span class="token keyword">const</span> options <span class="token operator">=</span> <span class="token punctuation">[</span>
  <span class="token punctuation">{</span> label<span class="token operator">:</span> <span class="token string">'Apple'</span> <span class="token punctuation">}</span><span class="token punctuation">,</span>
  <span class="token punctuation">{</span> label<span class="token operator">:</span> <span class="token string">'Orange'</span> <span class="token punctuation">}</span><span class="token punctuation">,</span>
  <span class="token punctuation">{</span> label<span class="token operator">:</span> <span class="token string">'Lemon'</span> <span class="token punctuation">}</span><span class="token punctuation">,</span>
  <span class="token punctuation">{</span> label<span class="token operator">:</span> <span class="token string">'Raspberry'</span> <span class="token punctuation">}</span><span class="token punctuation">,</span>
  <span class="token punctuation">{</span> label<span class="token operator">:</span> <span class="token string">'Strawberry'</span> <span class="token punctuation">}</span><span class="token punctuation">,</span>
<span class="token punctuation">]</span><span class="token punctuation">;</span>

<span class="token keyword">export</span> <span class="token keyword">function</span> <span class="token function">Example</span><span class="token punctuation">(</span><span class="token punctuation">)</span> <span class="token punctuation">{</span>
  <span class="token keyword">const</span> <span class="token punctuation">[</span>value<span class="token punctuation">,</span> setValue<span class="token punctuation">]</span> <span class="token operator">=</span> <span class="token function">useState</span><span class="token punctuation">(</span><span class="token string">'Lemon'</span><span class="token punctuation">)</span><span class="token punctuation">;</span>
  <span class="token keyword">return</span> <span class="token punctuation">(</span>
    <span class="token operator">&lt;</span><span class="token operator">></span>

      <span class="token operator">&lt;</span>label htmlFor<span class="token operator">=</span><span class="token string">"select"</span><span class="token operator">></span>
        Select
      <span class="token operator">&lt;</span><span class="token operator">/</span>label<span class="token operator">></span>
      <span class="token operator">&lt;</span>Select
        id<span class="token operator">=</span><span class="token string">"select"</span>
        value<span class="token operator">=</span><span class="token punctuation">{</span>value<span class="token punctuation">}</span>
        onValue<span class="token operator">=</span><span class="token punctuation">{</span>setValue<span class="token punctuation">}</span>
        options<span class="token operator">=</span><span class="token punctuation">{</span>options<span class="token punctuation">}</span>
      <span class="token operator">/</span><span class="token operator">></span>

      <span class="token operator">&lt;</span>label htmlFor<span class="token operator">=</span><span class="token string">"output"</span><span class="token operator">></span>
        Current value
      <span class="token operator">&lt;</span><span class="token operator">/</span>label<span class="token operator">></span>
      <span class="token operator">&lt;</span>output htmlFor<span class="token operator">=</span><span class="token string">"select"</span> id<span class="token operator">=</span><span class="token string">"output"</span><span class="token operator">></span>
        <span class="token punctuation">{</span><span class="token constant">JSON</span><span class="token punctuation">.</span><span class="token function">stringify</span><span class="token punctuation">(</span>value<span class="token punctuation">,</span> <span class="token keyword">undefined</span><span class="token punctuation">,</span> <span class="token string">' '</span><span class="token punctuation">)</span><span class="token punctuation">}</span>
      <span class="token operator">&lt;</span><span class="token operator">/</span>output<span class="token operator">></span>
    <span class="token operator">&lt;</span><span class="token operator">/</span><span class="token operator">></span>
  <span class="token punctuation">)</span><span class="token punctuation">;</span>
<span class="token punctuation">}</span>
</code></pre></main><script src="/react-listboxes/render.51b51511.js"></script></body></html>