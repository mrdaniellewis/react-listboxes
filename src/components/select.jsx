import React from 'react';
import PropTypes from 'prop-types';
import { options as validateOptions } from '../validators/options.js';
import { useNormalisedOptions } from '../hooks/use_normalised_options.js';
import { renderGroupedOptions } from '../helpers/render_grouped_options.js';

export function Select(rawProps) {
  const {
    options, setValue, blank: _1, value: _2, ...props
  } = useNormalisedOptions(rawProps);

  return (
    <select
      value={options.findIndex(o => o.selected)}
      onChange={({ target: { value: index } }) => setValue(options[+index]?.value ?? null)}
      {...props}
    >
      {renderGroupedOptions({
        options,
        renderGroup({ key, html, children, label }) { // eslint-disable-line react/prop-types
          return (
            <optgroup key={key} label={label} {...html}>
              {children}
            </optgroup>
          );
        },
        // eslint-disable-next-line react/prop-types
        renderOption({ label, key, node, html, disabled, index }) {
          return (
            <option
              value={index}
              id={key}
              key={key}
              disabled={disabled}
              {...html}
            >
              {node ?? label}
            </option>
          );
        },
      })}
    </select>
  );
}

Select.propTypes = {
  blank: PropTypes.node,
  setValue: PropTypes.func.isRequired,
  options: validateOptions.isRequired,
  value: PropTypes.any, // eslint-disable-line react/forbid-prop-types
};

Select.defaultProps = {
  blank: null,
  value: null,
};
