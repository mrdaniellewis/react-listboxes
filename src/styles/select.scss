@import "./variables.scss";

.select {
  display: inline-block;
  position: relative;

  > select {
    appearance: none;
    padding-right: 2em;
    border-radius: 2px;
    background: $color-button-background;
  }

  &::after {
    border: 0 solid $color-button-border;
    border-left-width: .5em;
    border-right-width: .5em;
    border-top-width: .5em;
    border-top-style: solid;
    border-top-color: $color-button-border;
    content: '';
    height: 0;
    position: absolute;
    pointer-events: none;
    right: .5em;
    transform: translateY(-50%);
    top: 50%;
    width: 0;
  }
}
