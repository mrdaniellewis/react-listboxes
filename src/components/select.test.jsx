import React, { useState } from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Select } from './select.jsx';

function SelectWrapper({ value: initialValue, ...props }) {
  const [value, setValue] = useState(initialValue);
  return (
    <Select value={value} setValue={setValue} {...props} />
  );
}

describe('options', () => {
  describe('as array of strings', () => {
    const options = ['Apple', 'Banana', 'Orange'];

    it('renders an select', () => {
      const { container } = render(
        <SelectWrapper options={options} />,
      );
      expect(container).toMatchSnapshot();
    });

    it('renders a select with a selected value', () => {
      const { container } = render(
        <SelectWrapper options={options} value="Orange" />,
      );
      const select = container.querySelector('select');
      expect(select).toHaveValue('2');
      expect(select.querySelector('option:nth-child(3)')).toHaveTextContent('Orange');
    });

    it('triggers the setValue callback with the selected value', () => {
      const spy = jest.fn();
      const { container } = render(<SelectWrapper options={options} setValue={spy} />);
      const select = container.querySelector('select');
      fireEvent.change(select, { target: { value: '1' } });
      expect(spy).toHaveBeenCalledWith('Banana');
    });

    it('updates when the value changes', () => {
      const { container } = render(<SelectWrapper options={options} />);
      const select = container.querySelector('select');
      fireEvent.change(select, { target: { value: '1' } });
      expect(select).toHaveValue('1');
      expect(select.querySelector('option:nth-child(2)')).toHaveTextContent('Banana');
    });
  });

  describe('options as array of numbers', () => {
    const options = [1, 2, 3];

    it('renders an select', () => {
      const { container } = render(
        <SelectWrapper options={options} />,
      );
      expect(container).toMatchSnapshot();
    });

    it('triggers the setValue callback with the selected value', () => {
      const spy = jest.fn();
      const { container } = render(<SelectWrapper options={options} setValue={spy} />);
      const select = container.querySelector('select');
      fireEvent.change(select, { target: { value: '1' } });
      expect(spy).toHaveBeenCalledWith(2);
    });

    it('updates when the value changes', () => {
      const { container } = render(<SelectWrapper options={options} />);
      const select = container.querySelector('select');
      fireEvent.change(select, { target: { value: '1' } });
      expect(select).toHaveValue('1');
      expect(select.querySelector('option:nth-child(2)')).toHaveTextContent('2');
    });
  });

  describe('options as array of objects', () => {
    describe('label', () => {
      const options = [{ label: 'Apple' }, { label: 'Banana' }, { label: 'Orange' }];

      it('renders an select', () => {
        const { container } = render(
          <SelectWrapper options={options} />,
        );
        expect(container).toMatchSnapshot();
      });

      it('triggers the setValue callback with the selected value', () => {
        const spy = jest.fn();
        const { container } = render(<SelectWrapper options={options} setValue={spy} />);
        const select = container.querySelector('select');
        fireEvent.change(select, { target: { value: '1' } });
        expect(spy).toHaveBeenCalledWith({ label: 'Banana' });
      });

      it('updates when the value changes', () => {
        const { container } = render(<SelectWrapper options={options} />);
        const select = container.querySelector('select');
        fireEvent.change(select, { target: { value: '1' } });
        expect(select).toHaveValue('1');
        expect(select.querySelector('option:nth-child(2)')).toHaveTextContent('Banana');
      });
    });

    describe('value', () => {
      it('is used as a options identity', () => {
        const options = [{ label: 'foo', value: 1 }, { label: 'foo', value: 2 }, { label: 'foo', value: 3 }];
        const spy = jest.fn();
        const { container } = render(<SelectWrapper options={options} value={2} setValue={spy} />);
        const select = container.querySelector('select');
        expect(select).toHaveValue('1');
        fireEvent.change(select, { target: { value: '2' } });
        expect(spy).toHaveBeenCalledWith({ label: 'foo', value: 3 });
      });
    });

    describe('id', () => {
      it('is used as a options identity', () => {
        const options = [{ label: 'foo', id: 1 }, { label: 'foo', id: 2 }, { label: 'foo', id: 3 }];
        const spy = jest.fn();
        const { container } = render(<SelectWrapper options={options} value={2} setValue={spy} />);
        const select = container.querySelector('select');
        expect(select).toHaveValue('1');
        fireEvent.change(select, { target: { value: '2' } });
        expect(spy).toHaveBeenCalledWith({ label: 'foo', id: 3 });
      });
    });

    describe('disabled', () => {
      it('sets the disabled attribute', () => {
        const options = [{ label: 'foo', disabled: true }];
        const { container } = render(<SelectWrapper options={options} />);
        const option = container.querySelector('option');
        expect(option).toHaveAttribute('disabled', '');
      });
    });

    describe('html', () => {
      it('sets attributes on the element', () => {
        const options = [{ label: 'foo', html: { 'data-foo': 'bar', className: 'class' } }];
        const { container } = render(<SelectWrapper options={options} />);
        const option = container.querySelector('option');
        expect(option).toHaveAttribute('data-foo', 'bar');
        expect(option).toHaveAttribute('class', 'class');
      });
    });

    describe('group', () => {
      const options = [
        { label: 'Apple' },
        { label: 'Orange', group: 'Citrus' },
        { label: 'Lemon', group: 'Citrus' },
        { label: 'Raspberry', group: 'Berry' },
        { label: 'Strawberry', group: 'Berry' },
      ];

      it('renders grouped options', () => {
        const { container } = render(<SelectWrapper options={options} />);
        expect(container).toMatchSnapshot();
      });

      it('triggers the setValue callback with the selected value', () => {
        const spy = jest.fn();
        const { container } = render(<SelectWrapper options={options} setValue={spy} />);
        const select = container.querySelector('select');
        fireEvent.change(select, { target: { value: '3' } });
        expect(spy).toHaveBeenCalledWith({ label: 'Lemon', group: 'Citrus' });
      });

      it('updates when the value changes', () => {
        const { container } = render(<SelectWrapper options={options} />);
        const select = container.querySelector('select');
        fireEvent.change(select, { target: { value: '3' } });
        expect(select).toHaveValue('3');
        expect(select.querySelector('optgroup:first-of-type > option:nth-child(2)')).toHaveTextContent('Lemon');
      });
    });

    describe('other attributes', () => {
      it('does not render them', () => {
        const options = [{ label: 'foo', 'data-foo': 'bar' }];
        const { container } = render(<SelectWrapper options={options} />);
        const option = container.querySelector('option');
        expect(option).not.toHaveAttribute('data-foo', 'bar');
      });
    });
  });

  describe('mapOption', () => {
    const options = [{ name: 'Apple' }, { name: 'Banana' }, { name: 'Orange' }];

    it('maps options', () => {
      const spy = jest.fn();
      const { container } = render(<SelectWrapper
        options={options}
        setValue={spy}
        mapOption={({ name }) => ({ label: name })}
      />);
      const select = container.querySelector('select');
      fireEvent.change(select, { target: { value: '2' } });
      expect(spy).toHaveBeenCalledWith({ name: 'Orange' });
    });

    it('selects a mapped option', () => {
      const { container } = render(<SelectWrapper
        options={options}
        value={{ name: 'Banana' }}
        mapOption={({ name }) => ({ label: name })}
      />);
      const select = container.querySelector('select');
      expect(select).toHaveValue('1');
      fireEvent.change(select, { target: { value: '2' } });
      expect(select.querySelector('option:nth-child(2)')).toHaveTextContent('Banana');
    });
  });
});

describe('blank', () => {
  const options = ['Apple', 'Banana', 'Orange'];

  it('renders a blank option', () => {
    const { container } = render(<SelectWrapper options={options} blank="Please select…" />);
    const select = container.querySelector('select');
    expect(select.querySelector('option')).toHaveTextContent('Please select…');
    expect(select).toHaveValue('0');
  });

  it('renders with a selected value', () => {
    const { container } = render(<SelectWrapper options={options} blank="Please select…" value="Orange" />);
    const select = container.querySelector('select');
    expect(select).toHaveValue('3');
    expect(select.querySelector('option:nth-child(4)')).toHaveTextContent('Orange');
  });
});

describe('OptionComponent', () => {
  const options = ['Apple', 'Banana', 'Orange'];

  it('allows additional props to be added to all options', () => {
    const { container } = render(
      <SelectWrapper options={options} OptionComponent={{ 'data-foo': 'bar' }} />,
    );
    container.querySelectorAll('option').forEach((option) => {
      expect(option).toHaveAttribute('data-foo', 'bar');
    });
  });

  it('allows the option component to be replaced', () => {
    function TestOption({ props }) {
      return (
        <option
          data-foo="bar"
          {...props}
        />
      );
    }
    const { container } = render(
      <SelectWrapper options={options} OptionComponent={TestOption} />,
    );
    container.querySelectorAll('option').forEach((option) => {
      expect(option).toHaveAttribute('data-foo', 'bar');
    });
  });
});

describe('OptGroupComponent', () => {
  const options = [
    { label: 'Apple' },
    { label: 'Orange', group: 'Citrus' },
    { label: 'Lemon', group: 'Citrus' },
    { label: 'Raspberry', group: 'Berry' },
    { label: 'Strawberry', group: 'Berry' },
  ];

  it('allows additional props to be added to all options', () => {
    const { container } = render(
      <SelectWrapper options={options} OptGroupComponent={{ 'data-foo': 'bar' }} />,
    );
    container.querySelectorAll('optgroup').forEach((option) => {
      expect(option).toHaveAttribute('data-foo', 'bar');
    });
  });

  it('allows the option component to be replaced', () => {
    function TestOptGroup({ props }) {
      return (
        <optgroup
          data-foo="bar"
          {...props}
        />
      );
    }
    const { container } = render(
      <SelectWrapper options={options} OptGroupComponent={TestOptGroup} />,
    );
    container.querySelectorAll('optgroup').forEach((option) => {
      expect(option).toHaveAttribute('data-foo', 'bar');
    });
  });
});

describe('additional props', () => {
  const options = ['Apple', 'Banana', 'Orange'];

  it('includes them on the select', () => {
    const { container } = render(
      <SelectWrapper options={options} required data-foo="bar" />,
    );
    const select = container.querySelector('select');
    expect(select).toHaveAttribute('required', '');
    expect(select).toHaveAttribute('data-foo', 'bar');
  });
});
