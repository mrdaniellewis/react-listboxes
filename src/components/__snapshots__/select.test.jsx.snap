// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options as array of strings renders an select 1`] = `
<div>
  <select>
    <option
      value="0"
    >
      Apple
    </option>
    <option
      value="1"
    >
      Banana
    </option>
    <option
      value="2"
    >
      Orange
    </option>
  </select>
</div>
`;

exports[`options options as array of numbers renders an select 1`] = `
<div>
  <select>
    <option
      value="0"
    >
      1
    </option>
    <option
      value="1"
    >
      2
    </option>
    <option
      value="2"
    >
      3
    </option>
  </select>
</div>
`;

exports[`options options as array of objects group renders grouped options 1`] = `
<div>
  <select>
    <option
      value="0"
    >
      Apple
    </option>
    <optgroup
      label="Citrus"
    >
      <option
        value="2"
      >
        Orange
      </option>
      <option
        value="3"
      >
        Lemon
      </option>
    </optgroup>
    <optgroup
      label="Berry"
    >
      <option
        value="5"
      >
        Raspberry
      </option>
      <option
        value="6"
      >
        Strawberry
      </option>
    </optgroup>
  </select>
</div>
`;

exports[`options options as array of objects label renders an select 1`] = `
<div>
  <select>
    <option
      value="0"
    >
      Apple
    </option>
    <option
      value="1"
    >
      Banana
    </option>
    <option
      value="2"
    >
      Orange
    </option>
  </select>
</div>
`;
